scripts/fbi/hate_crime/download_publication_data.py

```diff
@@ -0,0 +1,363 @@
+hence code uses enumerate to assign the table number.
+years_with_no_table0 list help make this correction.
+"""
+_YEARS_WITH_NO_TABLE0 = ('2004', '2005', '2006', '2007')
+
+# dictionary to access status across method calls
+# TODO(issue id here) remove this as global variable
+_STATUS_DICT = {
+    'year_list': [str(x) for x in range(2005, 2020)],
+    'output_dir': './source_data',
+    'cache_dir': './source_data/.cache_html/',
+    'force_fetch': True,
+    'base_url': 'https://ucr.fbi.gov/hate-crime',
+    'session': None,
+    'cur_year': '',
+    'cur_year_url': '',
+    'cur_access_url': '',
+    'table_html_url_list': [],
+    'download_urls': {}
+}
+
+
+def _get_abs_url(base_url: str, cur_url: str) -> str:
+    """Convert relative URL to absolute URL.
+  """
+    ret_url = cur_url
+    if not cur_url.startswith('http'):
+        if cur_url.startswith('/'):
+            ret_url = requests.compat.urljoin(base_url, cur_url)
+        else:
+            ret_url = base_url + '/' + cur_url
+    return ret_url
+
+
+def request_html_soup(cur_session, url: str) -> tuple:
+    """Get html text and bs4 object of the html from URL.
+    Args:
+        cur_session: requests session object which should be used to make url call.
+        url: The URL to be fetched.
+    Returns:
+        A tuple:
+            resp_html: The response HTML page text version.
+            resp_soup: BeautifulSoup object of the response HTML.
+  """
+    req = cur_session.get(url)
+    if req.status_code == requests.codes.ok:
+        resp_html = req.text
+        resp_soup = BeautifulSoup(resp_html, features="html.parser")
+    else:
+        resp_html = None
+        resp_soup = None
+
+        print('HTTP status code: ' + str(req.status_code))
+    return resp_html, resp_soup
+
+
+def _find_year_url(html_soup, year_str: str) -> str:
+    """Find the URL for a given year from the base HTML list of URL for all years.
+    Args:
+        html_soup: BeautifulSoup object of the HTML page with year URL.
+        year_str: string version of the year for which the URL is to be found.
+    Returns:
+        URL to the page for the data of particular year.
+  """
+    year_url = None
+    base_a_list = html_soup.find(id='content-core').find_all('a')
+    for cur_a in base_a_list:
+        cur_year = cur_a.string
+        if cur_year == year_str:
+            print(cur_year)
+            year_url = cur_a.get('href')
+    return year_url
+
+
+def _find_access_tables_url() -> str:
+    """Find the URL to 'Access Tables' in the given HTML for the year.
+  
+  The method uses status variables set in the status_dict for context.
+    Returns:
+        URL to the page with list of all the tables for a particular year.
+  """
+    global _STATUS_DICT
+    # fetch or return from cache
+    if _STATUS_DICT['force_fetch'] or not os.path.isfile(
+            os.path.join(_STATUS_DICT['cache_dir'],
+                         f"{_STATUS_DICT['cur_year']}.html")):
```

Comment:
`os.path.join(_STATUS_DICT['cache_dir'],  f"{_STATUS_DICT['cur_year']}.html")` is used multiple times.
Can set into a local var instead.

Reply:
fixed.